### PR TITLE
[SPARK-32083][SQL] Apply CoalesceShufflePartitions when input RDD has 0 partitions with AQE

### PR DESCRIPTION
**Pull request by manuzhang**
### What changes were proposed in this pull request?
As suggested by @cloud-fan in https://github.com/apache/spark/pull/28916#issuecomment-651527077, apply `CoalesceShufflePartitions` with partitionSpecs of `Nil` when `ShuffleQueryStageExec#mapStats` is None.

### Why are the changes needed?

For SQL like

```
SELECT b, COUNT(t1.a) as cnt
FROM t1
INNER JOIN t2
ON t1.id = t2.id
WHERE t1.id > 10
GROUP BY b
```

when all ids of t1 are smaller than 10. Many unnecessary tasks are launched for the final shuffle stage because `CoalesceShufflePartitions` is skipped when input RDD has 0 partitions.

#### Before 
image

#### After
image

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Updated tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala [hunk 1]
```diff
@@ -79,8 +66,29 @@ case class CoalesceShufflePartitions(session: SparkSession) extends Rule[SparkPl
           case stage: ShuffleQueryStageExec if stageIds.contains(stage.id) =>
             CustomShuffleReaderExec(stage, partitionSpecs)
         }
+      }
+
+      if (validMetrics.isEmpty) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala:71]: if a query stage has multiple leaf shuffles, and only one of them has 0-partition input RDD. What shall we do?

**Comment by cloud-fan** [comment]
cc @maryannxue @JkSelf @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala [hunk 2]
```diff
@@ -79,8 +66,29 @@ case class CoalesceShufflePartitions(session: SparkSession) extends Rule[SparkPl
...
+        updatePlan(Nil)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala:72]: Can you add a comment for it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala (hunk 1 above)
Comment (manuzhang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala:71]: I think it's like coalescing one less shuffles and handled by the `nonEmpty` codes.

**Comment by cloud-fan** [comment]
After more thoughts, maybe a better way is to add a new rule in `AdaptiveSparkPlanExec.optimizer`, which converts `LogicalQueryStage` to empty `LocalRelation` if the size is 0.

**Comment by JkSelf** [comment]
@manuzhang  I run the `test("Empty stage coalesced to 0-partition RDD")` in `AdaptiveQueryExecSuite`. It seems there is no unnecessary  tasks for the empty partitions. The Spark UI is as following.  And can you give a simple examples to reproduce this issue? Thanks.
image


The related code is :
```
spark.conf.set("spark.sql.adaptive.enabled", "true")
spark.conf.set("spark.sql.autoBroadcastJoinThreshold", "-1")
val df1 = spark.range(10).withColumn("a", 'id)
val df2 = spark.range(10).withColumn("b", 'id)
val result = df1.where('a > 10).join(df2.where('b > 10), "id").groupBy('a).count()
result.show()
```

**Comment by manuzhang** [comment]
@JkSelf Try changing `result.show()` to `result.collect()`

**Comment by JkSelf** [comment]
@manuzhang  It seems still only one stage and no unnecessary task for empty partitions.
image

related code:
```
spark.conf.set("spark.sql.adaptive.enabled", "true")
spark.conf.set("spark.sql.autoBroadcastJoinThreshold", "-1")
val df1 = spark.range(10).withColumn("a", 'id)
val df2 = spark.range(10).withColumn("b", 'id)
val result = df1.where('a > 10).join(df2.where('b > 10), "id").groupBy('a).count()
result.collect()
```

**Comment by manuzhang** [comment]
@JkSelf Could you check the `jobs` tab ?

**Comment by JkSelf** [comment]
@manuzhang  Here is the `jobs `tab.
image

**Comment by viirya** [comment]
When input RDD has 0 partitions, even `CoalesceShufflePartitions` is skipped, why we will launch unnecessary tasks?

**Comment by manuzhang** [comment]
@JkSelf @viirya 
Here is the partial SQL UI of running the same example with default number of shuffle partitions. (the binary is built from master branch till June 29). You can `SortMergeJoin` is followed by an `Exchange` of 200 partitions. 

image

**Comment by manuzhang** [comment]
@JkSelf 
I get the same result as you when I set the `numPartitions` of source to 1 (By default, it's 16 on my Mac), i.e.

```scala
val df1 = spark.range(0, 10, 1, 1).withColumn("a", 'id)
val df2 = spark.range(0, 10, 1, 1).withColumn("b", 'id)
val testDf = df1.where('a > 10).join(df2.where('b > 10), "id").groupBy('a).count()
testDf.collect()
```

Compare the execution plan with above.
image

**Comment by manuzhang** [comment]
@cloud-fan I find an issue with updating metrics if I convert a `LogicalQueryStage` to `LocalRelation` when its child's `mapStats` is empty. 

```
java.util.NoSuchElementException: key not found: 514
	at scala.collection.immutable.Map$Map1.apply(Map.scala:114)
	at org.apache.spark.sql.execution.ui.SQLAppStatusListener.$anonfun$aggregateMetrics$11(SQLAppStatusListener.scala:257)
```

The missing key is from the metrics of a `RangeExec` sent to `SQLAppStatusListener` earlier

**Pull request closed without merging**